Added auto-merging+branch deletion for dependabot, Dependency Reviewer, SonarLint, Snyk, and Docker

WilsonSue opened the pull request:
I've added multiple new actions and some docker files. How the auto-merge works is once the pull request is approved from dependabot, the change will be pushed onto main and the dependabot branch will be deleted. NOTE, please do not approve unless you are sure as you cannot undo the action and the checks have to pass before approving. I've also added sonarLint as a recommended vscode extension. I've added Docker files based on the requirements.

The dependabot auto merge and deletion has been tested in my test repo.

If you wish to run the python-package docker, please do so from the base directory. Alternatively, with the vscode-extension please do so in the vscode-extension directory with command

```shell
docker build -t my-node-app .
```

Side note, the check is not passing for this is due to the dependency issue detected by snyk. I haven't changed it because I'm not sure if you.

Please let me know if you notice any issues.

tgrigoruk commented:
There seems to be a lot going on here, this probably could have been several smaller PRs. The PR description could use more detail. Specifically:
- What is the purpose of this PR? (Why was it needed?)
- What are the "multiple new actions" added? Why were they added?
- What's all the `docker` stuff about? I see some commands provided but how/why are they used exactly?
- What the heck is `snyk`?
- Why aren't checks passing? (explanation given is incomplete) Is this not a problem??
- How does one test this PR?

This illustrates the benefit of using a PR template, as it reminds one to cover all the requisite bases.

WilsonSue commented:
# Description

This pull request includes the following changes:
- Added steps to upload coverage reports to Codecov.
- Integrated Snyk for vulnerability checking.
- Added a dependency review step for pull requests.
- Implemented auto-merge functionality for approved Dependabot PRs.
- Updated Poetry from version 1.8.2 to 1.8.3.
- Added SonarLint as a recommended extension for linting.

These changes aim to improve code quality and security by integrating Codecov for coverage reports and Snyk for vulnerability checks. Additionally, the auto-merge functionality streamlines dependency updates and deletes the Dependabot branch.

Fixes # (issue)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

These changes have been tested on a test repository with Dependabot. The PRs were approved and merged successfully. They are resulting in the Dependabot branches being deleted as expected. However, there's a bug where if a lint check or dependency check does not pass, the auto merge and branch deletion cannot occur.

# How to test the auto merge and branch deletion for Dependabot?

Approve a pull request from Dependabot

# Description of what Snyk is and how it works

Snyk is a developer-first security tool that automatically finds and fixes vulnerabilities in your dependencies. It integrates with your development workflow to provide continuous monitoring and security fixes for your projects.

# Description of what Codecov is and how it works

Codecov is a tool that provides metrics on how well your code is tested. It integrates with CI pipelines to report code coverage and helps identify untested parts of your codebase. It supports various coverage formats and provides insightful reports to improve code quality.

# Updated Poetry from 1.8.2 to 1.8.3

The Poetry package manager has been updated to version 1.8.3. This update includes various improvements and bug fixes that enhance dependency management. Also, it fixes the Snyk warning check

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
